### PR TITLE
[NT-1453][NT-1344][NT-1399][NT-1509]: Fix Pledge with Digital addOns 

### PR DESCRIPTION
**Pull request by Arkariang**
# 📲 What

- Pledge to a Digital Reward with Digital AddOns

# 🛠 How

- Refactor in PledgeFragmentViewModel
- Extracted shipping cost (Reward || Reward + AddOns)
- Extracted pledgeAmount ( Reward || Rewards + AddOns || NoReward )
- Extracted cost for DigitalReward
- Updated MaxPledge -> MaxPerProject - current pledgeAmount

# 👀 See

PledgeDigitalAddOns

|  |  |

# 📋 QA

- Pledge to a Reward no Reward -> Update that pledge -> Change payment method for that pledge
- Pledge to a Digital Reward -> Update that pledge -> Change payment method for that pledge
- Pledge to a Reward with limited shipping -> Update that pledge -> Change payment method for that pledge
- Pledge to a Reward with unrestricted shipping -> Update that pledge -> Change payment method for that pledge

- Pledge to a Digital Reward + AddOns -> Update that pledge -> Change payment method for that pledge
- Pledge to a Reward with limited shipping + AddOns -> Update that pledge -> Change payment method for that pledge
- Pledge to a Reward with unrestricted shipping + AddOns -> Update that pledge -> Change payment method for that pledge

# Story 📖

[NT-1344](https://kickstarter.atlassian.net/browse/NT-1344)
[NT-1399](https://kickstarter.atlassian.net/browse/NT-1399)
[NT-1453](https://kickstarter.atlassian.net/browse/NT-1453)
[NT-1509](https://kickstarter.atlassian.net/browse/NT-1509)

### CODE REVIEW

**Comment by jgsamudio** [comment]
Don't know if this is a known issue, but when I select the no reward option for the one inch straw project, I noticed that the default is 1 MX where it should be 10 MX. When I increment the bonus support it increments it by 10 instead of by 1

File: app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt [hunk 1]
```diff
@@ -620,16 +640,25 @@ interface PledgeFragmentViewModel {
...
 
-            val backingAmount = backing
+            // - For no Reward the amount of the RW and the bonus amount are the same value
+            val backingAmountNR = backing
+                    .filter { it.reward() == null }
+                    .map { it.bonusAmount() }
+                    .distinctUntilChanged()
+
+            val backingAmountRW = backing
+                    .filter { it.reward()?.let { rw -> !RewardUtils.isNoReward(rw) }?: false }
```
Comment (jgsamudio) [inline app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt:655]:
```suggestion
                    .filter { it.reward()?.let { rw -> !RewardUtils.isNoReward(rw) } ?: false }
```

File: app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt [hunk 2]
```diff
@@ -738,26 +767,24 @@ interface PledgeFragmentViewModel {
             val backingShippingAmount = backing
                     .map { it.shippingAmount() }
 
-            // - Shipping amount when backing
-            Observable.combineLatest(this.shippingRule, pledgeReason, backingShippingAmount, rewardAndAddOns) {rule, reason, bShippingAmount, listRw ->
+            val shippingAmountWhenBacking = Observable.combineLatest(this.shippingRule, pledgeReason, backingShippingAmount, rewardAndAddOns) {rule, reason, bShippingAmount, listRw ->
```
Comment (jgsamudio) [inline app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt:770]:
```suggestion
            val shippingAmountWhenBacking = Observable.combineLatest(this.shippingRule, pledgeReason, backingShippingAmount, rewardAndAddOns) { rule, reason, bShippingAmount, listRw ->
```

File: app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt [hunk 3]
```diff
@@ -738,26 +767,24 @@ interface PledgeFragmentViewModel {
...
-                        this.shippingAmount.onNext(ProjectViewUtils.styleCurrency(it.first, it.second, this.ksCurrency))
-                    }
 
             val newPledge = pledgeReason
                     .filter { it == PledgeReason.PLEDGE || it == PledgeReason.UPDATE_REWARD }
                     .map { it }
 
             // - Shipping amount when no backing
-            Observable.combineLatest(this.shippingRule, newPledge, rewardAndAddOns) {rule, reason, listRw ->
+            val shippingAmountNewPledge = Observable.combineLatest(this.shippingRule, newPledge, rewardAndAddOns) {rule, reason, listRw ->
```
Comment (jgsamudio) [inline app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt:780]:
```suggestion
            val shippingAmountNewPledge = Observable.combineLatest(this.shippingRule, newPledge, rewardAndAddOns) { rule, reason, listRw ->
```

File: app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt [hunk 4]
```diff
@@ -775,33 +802,35 @@ interface PledgeFragmentViewModel {
                         this.shippingRulesSectionIsGone.onNext(true )
                     }
 
+            val isRewardWithShipping = this.selectedReward
+                    .filter { RewardUtils.isShippable(it) }
+                    .map { it }
```
Comment (jgsamudio) [inline app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt:807]:
```suggestion
```

File: app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt [hunk 5]
```diff
@@ -775,33 +802,35 @@ interface PledgeFragmentViewModel {
...
+
+            val isDigitalRw = this.selectedReward
+                    .filter { RewardUtils.isDigital(it) }
+                    .map { it }
```
Comment (jgsamudio) [inline app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt:811]:
```suggestion
```

File: app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt (hunk 4 above)
Comment (jgsamudio) [reply app/src/main/java/com/kickstarter/viewmodels/PledgeFragmentViewModel.kt:807]: No need for this line as its mapping to itself

**Review by jgsamudio — CHANGES_REQUESTED** [review]

**Comment by jgsamudio** [comment]
When selecting a reward in the "one inch straw" project, I noticed that the rewards headers scrolls because the price is moved to a multiple line. Is this related to this PR or should we create a ticket for this?

Screen Shot 2020-09-08 at 7 27 10 PM

**Comment by Arkariang** [comment]
> Don't know if this is a known issue, but when I select the no reward option for the one inch straw project, I noticed that the default is 1 MX where it should be 10 MX. When I increment the bonus support it increments it by 10 instead of by 1

Will be done here https://kickstarter.atlassian.net/browse/NT-1516

**Comment by Arkariang** [comment]
> When selecting a reward in the "one inch straw" project, I noticed that the rewards headers scrolls because the price is moved to a multiple line. Is this related to this PR or should we create a ticket for this?
> 
> Screen Shot 2020-09-08 at 7 27 10 PM

will be done here https://kickstarter.atlassian.net/browse/NT-1512

**Review by jgsamudio — APPROVED** [review]

**Pull request merged**